### PR TITLE
Ii 9587 servicenow agent

### PR DESCRIPTION
**Pull request by hmaccelerate**
finished the implementation of servicenow agent

### CODE REVIEW

**Review by zhangzinan — COMMENTED** [review]
1. please also edit `.CONFIGVARS.md` and `README.md` if you change `config.ini.template`
2. please use `../utils/make-agent-installer.sh` to build `servicenow.tar.gz`

**Review by crespo5609 — APPROVED** [review]

**Comment by dearmore** [comment]
Historically, the config.ini is a static file from the agent's perspective configured by the user.

There are two different flows through the agent file:
1) historical data -- manually run the python script by the user
2) current data -- run by cron on a continuing basis

1) This flow should be updated in the agent as this agent currently modifies the config.ini, additionally, the current time variable should be moved out of the config file and into a status file.

2) Addi a cache (local sqlite3 db) to the agent to configure mapping of an instance to an alias: mariadb agent for reference. 

3) Update isHistorical to use true/false instead of 0/1

File: servicenow/getlogs_servicenow-3.8.py [hunk 1]
```diff
@@ -0,0 +1,1644 @@
...
+    # get localized earliest datetime
+    local_earliest_datetime = datetime.fromtimestamp(utc_earliest_epoch)
+    # convert earliest datetime to the data timezone
+    data_earliest_datetime = local_earliest_datetime.astimezone(
+        agent_config_vars['timezone'])
+    # do not apply timezone conversion later
+    # agent_config_vars['timezone'] = pytz.utc
+    # convert to string for Glide
+    earliest_date_and_time = data_earliest_datetime.strftime(
+        agent_config_vars['timestamp_format'][0]).split(' ')
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:57]: The condition for this code is if the historical flag is not set and there is no status file with a cron start time.  We should move it to the related section of code (line 83).

File: servicenow/getlogs_servicenow-3.8.py [hunk 2]
```diff
@@ -0,0 +1,1644 @@
...
+    ## add timestamp query
+    # get utc earliest datetime
+    utc_earliest_epoch = time.time()
+    # get localized earliest datetime
+    local_earliest_datetime = datetime.fromtimestamp(utc_earliest_epoch)
+    # convert earliest datetime to the data timezone
+    data_earliest_datetime = local_earliest_datetime.astimezone(
+        agent_config_vars['timezone'])
+    # do not apply timezone conversion later
+    # agent_config_vars['timezone'] = pytz.utc
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:54]: If this code is not needed, we should remove it.

File: servicenow/getlogs_servicenow-3.8.py [hunk 3]
```diff
@@ -0,0 +1,1644 @@
...
+    elif agent_config_vars['is_historical'] == False:
+        if agent_config_vars['cron_start_time']:
+            utc_cron_epoch = agent_config_vars['cron_start_time']
+            # get localized earliest datetime
+            local_cron_datetime = datetime.fromtimestamp(float(utc_cron_epoch))
+            # convert earliest datetime to the data timezone
+            data_cron_datetime = local_cron_datetime.astimezone(
+                agent_config_vars['timezone'])
+            # do not apply timezone conversion later
+            agent_config_vars['timezone'] = pytz.utc
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:76]: Why do we have this commented out if the cron_start_time is missing but included if the variable is present. Also why is this hard coded to utc?

File: servicenow/getlogs_servicenow-3.8.py [hunk 4]
```diff
@@ -0,0 +1,1644 @@
...
+    for i in range(minlen(fields, message)):
+        json[fields[i]] = message[i]
+    return json
+
+
+def minlen(one, two):
+    return min(len(one), len(two))
+
+
+def parse_csv_message(message):
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:1021]: It doesn't look like the parse_csv_message method is used at all.  Do we need to keep it? If not, can we remove the parse_csv_data and the parse_csv_row methods too?

File: servicenow/getlogs_servicenow-3.8.py [hunk 5]
```diff
@@ -0,0 +1,1644 @@
...
+        for filter_value in filter_values.split(','):
+            statement = '{keyword}*{value}^{keyword}ISNOTEMPTY'.format(keyword=filter_keyword, value=filter_value)
+            # OR between values
+            filter_statement = '{}^OR{}'.format(filter_statement, statement) if len(
+                filter_statement) != 0 else statement
+        # AND between keywords
+        passthru['sysparm_query'] = '{}^{}'.format(passthru['sysparm_query'], filter_statement) if len(
+            passthru['sysparm_query']) != 0 else '{}'.format(filter_statement)
+    # clear since it's handled already
+    agent_config_vars['filters_include'] = ''
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:108]: Do we need to clear this variable? Typically, the agent_config_vars is a static variable reflecting the configuration in the config.ini.

File: servicenow/getlogs_servicenow-3.8.py [hunk 6]
```diff
@@ -0,0 +1,1644 @@
...
+        for filter_value in filter_values.split(','):
+            statement = '{keyword}!*{value}^{keyword}ISNOTEMPTY'.format(keyword=filter_keyword, value=filter_value)
+            # OR between values
+            filter_statement = '{}^OR{}'.format(filter_statement, statement) if len(
+                filter_statement) != 0 else statement
+        # AND between keywords
+        passthru['sysparm_query'] = '{}^{}'.format(passthru['sysparm_query'], filter_statement) if len(
+            passthru['sysparm_query']) != 0 else '{}'.format(filter_statement)
+    # clear since it's handled already
+    agent_config_vars['filters_exclude'] = ''
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:121]: Do we need to clear this variable? Typically, the agent_config_vars is a static variable reflecting the configuration in the config.ini.

File: servicenow/getlogs_servicenow-3.8.py [hunk 7]
```diff
@@ -0,0 +1,1644 @@
...
+    passthru['sysparm_query'] = '{}^{}'.format(passthru['sysparm_query'], agent_config_vars['addl_query']) if len(
+        agent_config_vars['addl_query']) != 0 else passthru['sysparm_query']
+    # build auth
+    auth = (agent_config_vars['username'], ifobfuscate.decode(agent_config_vars['password']))
+    # logger.debug(ifobfuscate.decode(agent_config_vars['password']))
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:126]: Remove commented debugging code

File: servicenow/getlogs_servicenow-3.8.py [hunk 8]
```diff
@@ -0,0 +1,1644 @@
...
+            instance_field = config_parser.get('agent', 'instance_field', raw=True)
+            device_field = config_parser.get('agent', 'device_field', raw=True)
+            timestamp_field = config_parser.get('agent', 'timestamp_field', raw=True) or 'timestamp'
+            timestamp_format = config_parser.get('agent', 'timestamp_format', raw=True) or 'epoch'
+            timezone = config_parser.get('agent', 'timezone') or 'UTC'
+            data_fields = config_parser.get('agent', 'data_fields', raw=True)
+            start_time = config_parser.get('agent', 'start_time', raw=True)
+            end_time = config_parser.get('agent', 'end_time', raw=True)
+            is_historical = config_parser.get('agent', 'is_historical', raw=True)
+            # cron_start_time = config_parser.get('agent', 'cron_start_time', raw=True)
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:195]: Remove deprecated code

File: servicenow/getlogs_servicenow-3.8.py [hunk 9]
```diff
@@ -0,0 +1,1644 @@
...
+            agent_http_proxy = config_parser.get('agent', 'agent_http_proxy')
+            agent_https_proxy = config_parser.get('agent', 'agent_https_proxy')
+
+            # filters
+            filters_include = config_parser.get('agent', 'filters_include')
+            filters_exclude = config_parser.get('agent', 'filters_exclude')
+
+            # message parsing
+            json_top_level = config_parser.get('agent', 'json_top_level')
+            # project_field = config_parser.get('agent', 'project_field', raw=True)
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:185]: Remove deprecated code

File: servicenow/getlogs_servicenow-3.8.py [hunk 10]
```diff
@@ -0,0 +1,1644 @@
...
+            agent_proxies['https'] = agent_https_proxy
+
+        # filters
+        if len(filters_include) != 0:
+            filters_include = filters_include.split('|')
+        if len(filters_exclude) != 0:
+            filters_exclude = filters_exclude.split('|')
+
+        # fields
+        # project_fields = project_field.split(',')
```
Comment (dearmore) [inline servicenow/getlogs_servicenow-3.8.py:229]: Remove commented code if not needed

**Review by dearmore — APPROVED** [review]

**Pull request merged**